core/src/mbc/mbc5.rs

```diff
@@ -0,0 +1,121 @@
+use crate::{
+    cartridge::{RamSize, RomSize},
+    primitives::{Byte, Word},
+};
+use super::Mbc;
+
+/// MBC5.
+///
+/// With this controller, the cartridge can have up to 8MiB of ROM and up to
+/// 128KiB of external RAM.
+pub(crate) struct Mbc5 {
+    rom: Box<[Byte]>,
+    ram: Box<[Byte]>,
+
+    /// A 9 bit number to select the bank mapped to 0x4000 -- 0x8000. Values 0
+    /// to 0x1E0. In MBC5 you can actually select bank 0 here to map bank 0
+    /// twice. Bits 9 to 15 are always 0.
+    rom_bank: u16,
+
+    /// A 4 bit number to select the RAM bank. Values 0 to 0xF.
+    ram_bank: u8,
+
+    /// Whether or not the RAM is enabled.
+    ram_enabled: bool,
+}
+
+
+impl Mbc5 {
+    pub(crate) fn new(data: &[u8], rom_size: RomSize, ram_size: RamSize) -> Self {
+        assert!(
+            rom_size.len() == data.len(),
+            "Length of cartridge doesn't match length specified in ROM size header",
+        );
+
+        let rom: Vec<_> = data.iter().cloned().map(Byte::new).collect();
+        let ram = vec![Byte::zero(); ram_size.len()];
```

Comment:
Here should be a check for the RAM size. Allowed sizes are `64kbit, 256kbit and 1mbit.` according to [[1]](http://gbdev.gg8.se/wiki/articles/MBC5). And I think our `RamSize` is currently limited to `RamSize::Kb32` this should be changed. When this is changed `Mbc1` and `Mbc3` should be limited to a ram size of `RamSize::Kb32` according to [[2]](http://gbdev.gg8.se/wiki/articles/MBC1) and [[3]](http://gbdev.gg8.se/wiki/articles/MBC3).

- \[1\]: http://gbdev.gg8.se/wiki/articles/MBC5
- \[2\]: http://gbdev.gg8.se/wiki/articles/MBC1
- \[3\]: http://gbdev.gg8.se/wiki/articles/MBC3